src/codegen/codegen_c_visitor.cpp

```diff
@@ -3252,88 +3185,122 @@ std::string CodegenCVisitor::get_range_var_float_type(const SymbolType& symbol)
     return float_data_type();
 }
 
-/**
- * \details For CPU/Host target there is no device pointer. In this case
- * just use the host variable name directly.
- */
-std::string CodegenCVisitor::get_variable_device_pointer(const std::string& variable,
-                                                         const std::string& /*type*/) const {
-    return variable;
-}
-
 
 void CodegenCVisitor::print_instance_variable_setup() {
     if (range_variable_setup_required()) {
         print_setup_range_variable();
     }
 
-    printer->add_newline(2);
+    printer->add_newline();
+    printer->add_line("// Allocate instance structure");
+    printer->fmt_start_block("static void {}(NrnThread* nt, Memb_list* ml, int type)",
+                             method_name(naming::NRN_PRIVATE_CONSTRUCTOR_METHOD));
+    printer->add_line("assert(!ml->instance);");
+    printer->add_line("assert(!ml->global_variables);");
+    printer->add_line("assert(ml->global_variables_size == 0);");
+    printer->fmt_line("auto* const inst = new {}{{}};", instance_struct());
+    printer->fmt_line("assert(inst->{} == &{});",
+                      naming::INST_GLOBAL_MEMBER,
+                      global_struct_instance());
+    printer->add_line("ml->instance = inst;");
+    printer->fmt_line("ml->global_variables = inst->{};", naming::INST_GLOBAL_MEMBER);
+    printer->fmt_line("ml->global_variables_size = sizeof({});", global_struct());
+    printer->end_block(2);
+
+    auto const cast_inst_and_assert_validity = [&]() {
+        printer->fmt_line("auto* const inst = static_cast<{}*>(ml->instance);", instance_struct());
+        printer->add_line("assert(inst);");
+        printer->fmt_line("assert(inst->{});", naming::INST_GLOBAL_MEMBER);
+        printer->fmt_line("assert(inst->{} == &{});",
+                          naming::INST_GLOBAL_MEMBER,
+                          global_struct_instance());
+        printer->fmt_line("assert(inst->{} == ml->global_variables);", naming::INST_GLOBAL_MEMBER);
+        printer->fmt_line("assert(ml->global_variables_size == sizeof({}));", global_struct());
+    };
+
+    printer->fmt_line(
+        "static inline void copy_instance_to_device(NrnThread* nt, Memb_list* ml, {} const* inst);",
+        instance_struct());
+    printer->fmt_line("static inline void delete_instance_from_device({}& inst);",
+                      instance_struct());
```

Comment:
No strong opinion but maybe we could remove completely `device` related functions from the `C` codegen?

Reply:
These are just empty definitions. This is then overridden in OpenACC visitor.

These are callback functions registered to coreneuron and hence they are defined in C visitor as well. Maybe we could register nullptr? But maybe they could be repurposed later for generic mechanism-related cleanup by just renaming.